src/core/electrostatics_magnetostatics/coulomb.cpp

```diff
@@ -0,0 +1,555 @@
+    else
+#endif
+      p3m_calc_kspace_forces(1, 0);
+    break;
+#endif
+  case COULOMB_MMM2D:
+    MMM2D_add_far_force();
+    MMM2D_dielectric_layers_force_contribution();
+    break;
+#ifdef SCAFACOS
+  case COULOMB_SCAFACOS:
+    assert(!Scafacos::dipolar());
+    Scafacos::add_long_range_force();
+    break;
+#endif
+  default:
+    break;
+  }
+}
+
+void energy_calc_long_range_coulomb_energy(Observable_stat &energy) {
+  switch (coulomb.method) {
+#ifdef P3M
+  case COULOMB_P3M_GPU:
+    printf(
+        "long range energy calculation not implemented for GPU P3M\n"); // TODO
+    // make
+    // right
+    break;
+  case COULOMB_P3M:
+    p3m_charge_assign();
+    energy.coulomb[1] = p3m_calc_kspace_forces(0, 1);
+    break;
+  case COULOMB_ELC_P3M:
+    // assign the original charges first
+    // they may not have been assigned yet
+    p3m_charge_assign();
+    if (!elc_params.dielectric_contrast_on)
+      energy.coulomb[1] = p3m_calc_kspace_forces(0, 1);
+    else {
+      energy.coulomb[1] = 0.5 * p3m_calc_kspace_forces(0, 1);
+      energy.coulomb[1] += 0.5 * ELC_P3M_dielectric_layers_energy_self();
+
+      //  assign both original and image charges now
+      ELC_p3m_charge_assign_both();
+      ELC_P3M_modify_p3m_sums_both();
+
+      energy.coulomb[1] += 0.5 * p3m_calc_kspace_forces(0, 1);
+
+      // assign only the image charges now
+      ELC_p3m_charge_assign_image();
+      ELC_P3M_modify_p3m_sums_image();
+
+      energy.coulomb[1] -= 0.5 * p3m_calc_kspace_forces(0, 1);
+    }
+    energy.coulomb[2] = ELC_energy();
+    break;
+#endif
+#ifdef SCAFACOS
+  case COULOMB_SCAFACOS:
+    assert(!Scafacos::dipolar());
+    energy.coulomb[1] += Scafacos::long_range_energy();
+    break;
+#endif
+  case COULOMB_MMM2D:
+    *energy.coulomb += MMM2D_far_energy();
+    *energy.coulomb += MMM2D_dielectric_layers_energy_contribution();
+    break;
+  default:
+    break;
+  }
+}
+
+void energy_n_coulomb(int &n_coulomb) {
+  switch (coulomb.method) {
+  case COULOMB_NONE:
+    n_coulomb = 0;
+    break;
+  case COULOMB_ELC_P3M:
+    n_coulomb = 3;
+    break;
+  case COULOMB_P3M_GPU:
+  case COULOMB_P3M:
+    n_coulomb = 2;
+    break;
+  case COULOMB_SCAFACOS:
+    n_coulomb = 2;
+    break;
+  default:
+    n_coulomb = 1;
+  }
+}
+
+void icc_calc_pair_coulomb_force(Particle *p1, Particle *p2, double *d,
+                                 double dist, double dist2, double *force) {
+  auto const q1q2 = p1->p.q * p2->p.q;
+  if (q1q2 != 0) {
+    switch (coulomb.method) {
+#ifdef P3M
+    case COULOMB_ELC_P3M:
+    case COULOMB_P3M_GPU:
+    case COULOMB_P3M:
+      p3m_add_pair_force(q1q2, d, dist2, dist, force);
+      break;
+#endif /* P3M */
+    case COULOMB_MMM1D:
+      add_mmm1d_coulomb_pair_force(q1q2, d, dist2, dist, force);
+      break;
+    case COULOMB_MMM2D:
+      add_mmm2d_coulomb_pair_force(q1q2, d, dist2, dist, force);
+      break;
+    default:
+      break;
+    }
+  }
+}
+
+void icc_calc_long_range_force() {
```

Comment:
This should be removed, icc should use the normal coulomb longrange function, the same for the pair forces